backup scheduler configuration options

panagiotisgts opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->
Introduce the configuration options for the backup scheduler in the unified configuration under `camunda.data.primary-storage.backup` and in the Zeebe broker configurations under `zeebe.broker.data.backupScheduler`.

The `required` property will be used to force validate the scheduler and the backup store config to block broker startup if misconfigured.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #40439

panagiotisgts committed:
feat: enable backup scheduler configuration params

lenaschoenburg commented on configuration/src/main/java/io/camunda/configuration/PrimaryStorageBackup.java:
```diff
@@ -0,0 +1,112 @@
+    this.continuous = continuous;
+  }
+
+  public String getSchedule() {
```
💭 We could detect the different schemes (cron vs interval) and return a matching type here.

```java
sealed interface Schedule {
  record Cron(SomeTypeFromCronLibrary expression) implements Schedule {}
  record Delay(Duration interval) implements Schedule {}
}
```

panagiotisgts committed:
feat: typed schedule retrieval

Allow scheduler expressions parsed as typed definitions

panagiotisgts committed:
deps: add cron dependency

panagiotisgts committed:
feat: support none and auto schedules

None and auto schedules should be supported to provide flexibility for retention clean up.

Auto schedule type allows for a custom duration supplier to be used for interval determination

panagiotisgts committed:
refactor: directly pass BackupSchedulerCfg to DataCfg

Unfortunately, we cannot have the configuration as part of the unified config module, as it forms a cyclic dependency. Instead, we pass it directly from the unified config to the datacfg during startup.

Force retention schedule configuration existence.

panagiotisgts committed:
feat: validate schedule configuration if enabled on systemcontext startup

panagiotisgts commented on zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/SystemContext.java:
```diff
@@ -563,6 +566,17 @@ private void validateListenersConfig(final GlobalListenersCfg listeners) {
   }
 
+  private void validateBackupSchedulerConfig(final BackupSchedulerCfg backupSchedulerCfg) {
+    if (!backupSchedulerCfg.isContinuous() || !backupSchedulerCfg.isRequired()) {
```
The `isRequired` is essentially a flag that would force validation regarding the scheduler config and also potentially the backup store config presence and validity.

panagiotisgts replied:
To essentially fail startup if any of the scheduler/backup store config is not correct

panagiotisgts commented on zeebe/backup/src/main/java/io/camunda/zeebe/backup/schedule/Schedule.java:
```diff
@@ -0,0 +1,167 @@
+   */
+  Optional<Instant> previousExecution(Instant from);
+
+  ScheduleType getType();
```
Mostly because I don't fancy type assertions that much :upside_down_face: The only time we would have to do a type check is when the schedule is set to `auto` where we'd do some calculation to provide the values I guess.

lenaschoenburg replied:
I'd still suggest removing this, a sealed interface is basically a built-in type enum that guarantees that the enum and the actual implementation types never drift apart.

panagiotisgts committed:
refactor: remove type enum

It'll be based only on the type assertion

panagiotisgts committed:
refactor: align schedules config after unified refactor

panagiotisgts committed:
fix: validate SystemContext for scheduler

lenaschoenburg reviewed (approved):
Nice work 🚀

panagiotisgts committed:
fix: emit the date parsing exception from schedule parsing
